File: license-server/src/format.js
// ── Talus license key format codec ─────────────────────────────────────────
//
// Canonical (internal) form — what the Ed25519 signature actually covers:
//     base64(JSON payload) "." base64(64-byte signature)
//
// Display (pretty) form — what customers receive, type, and read aloud:
//     TALUS-XXXXX-XXXXX-XXXXX-…-XXXXX-SS
//
//   • XXXXX = Crockford base32 (0-9, A-Z minus I, L, O, U) of the canonical
//     bytes, grouped in fives — no ambiguous characters, copy-paste safe.
//   • "SS"   = last two hex chars of SHA-256 over the canonical bytes, so
//     typos are caught client-side before activation.
//
// The signature is PART of the encoded bytes, so the pretty form carries the
// full signed key. Decoding is lossless; verification always operates on the
// canonical form. Old canonical keys remain accepted everywhere.

const ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
const PREFIX = 'TALUS';
const GROUP = 5;

export function is_canonical(key) {
  const k = String(key).trim();
  return k.includes('.') && /^[A-Za-z0-9+/=_-]+\.[A-Za-z0-9+/=_-]+$/.test(k);
}

export function is_pretty(key) {
  const k = String(key).trim().toUpperCase().replace(/\s+/g, '');
  // Many groups = full JWT encoded. Exactly 4 groups = SHORT key: not a
  // self-contained encoding — it resolves to the JWT via admin/bind.
  return (
    k.startsWith(PREFIX + '-') &&
    /^[0-9A-Z-]+$/.test(k) &&
    (k.match(/-/g) ?? []).length >= 6
  );
}

// ── SHORT keys: TALUS-XXXXX-XXXXX-XXXXX-XXXXX ──────────────────────────
//
// A short key is a random 104-bit handle (Crockford base32). The signed
// JWT (canonical form) is bound to it server-side (admin/bind) and the
// client fetches the JWT at activation, verifies the Ed25519 signature
// locally, and proceeds as before. The short key is stored ONLY as a
// SHA-256 hash — seeing the D1 database does not reveal usable keys.

export function is_short(key) {
  const k = String(key).trim().toUpperCase().replace(/\s+/g, '');
  return (
    k.startsWith(PREFIX + '-') &&
    /^[0-9A-Z-]+$/.test(k) &&
    (k.match(/-/g) ?? []).length === 4
  );
}

// Normalize a short key: trim, collapse whitespace, uppercase. Crockford
// alphabet has no I/L/O/U, so map common mistypes to their digits.
export function normalize_short_key(key) {
  const raw = String(key).trim().toUpperCase().replace(/\s+/g, '');
  if (!raw.startsWith(PREFIX + '-')) return raw;
  return (
    PREFIX +
    '-' +
    raw
      .slice(PREFIX.length + 1)
      .replace(/O/g, '0')
      .replace(/[IL]/g, '1')
  );
}

// Canonical key string → pretty key string (sha256_hex_fn: bytes → hex).
export async function to_pretty(canonical_key, sha256_hex_fn) {
  const data = new TextEncoder().encode(String(canonical_key).trim());
  const digest = await sha256_hex_fn(data);
  const body = b32_encode(data);
  const grouped = (body.match(new RegExp(`.{1,${GROUP}}`, 'g')) ?? []).join('-');
  return `${PREFIX}-${grouped}-${digest.slice(-2).toUpperCase()}`;
}

// Pretty key string → canonical key string. Throws on checksum mismatch.
// (sha256_hex_sync: bytes → hex string, e.g. via node:crypto.)
export function decode_pretty(pretty_key, sha256_hex_sync) {
  const k = String(pretty_key).trim().toUpperCase().replace(/\s+/g, '');
  if (!k.startsWith(PREFIX + '-')) throw new Error('not a pretty key');
  const parts = k.slice(PREFIX.length + 1).split('-');
  if (parts.length < 2) throw new Error('key too short');
  const checksum = parts[parts.length - 1];
  const body = parts.slice(0, -1).join('');
  if (!/^[0-9ABCDEFGHJKMNPQRSTVWXYZ]+$/.test(body)) {
    throw new Error('invalid characters in key');
  }
  const data = b32_decode(body);
  const expected = sha256_hex_sync(data).slice(-2).toUpperCase();
  if (checksum !== expected) throw new Error('checksum mismatch — retype the key');
  return new TextDecoder().decode(data);
}

// ── base32 (Crockford alphabet) helpers ────────────────────────────────────

function b32_encode(bytes) {
  let out = '';
  let acc = 0;
  let nbits = 0;
  for (const byte of bytes) {
    acc = (acc << 8) | byte;
    nbits += 8;
    while (nbits >= 5) {
      out += ALPHABET[(acc >>> (nbits - 5)) & 31];
      nbits -= 5;
    }
  }
  if (nbits > 0) out += ALPHABET[(acc << (5 - nbits)) & 31];
  return out;
}

function b32_decode(s) {
  const out = [];
  let acc = 0;
  let nbits = 0;
  for (const ch of s) {
    const idx = ALPHABET.indexOf(ch);
    if (idx < 0) throw new Error('invalid character');
    acc = (acc << 5) | idx;
    nbits += 5;
    if (nbits >= 8) {
      out.push((acc >>> (nbits - 8)) & 0xff);
      nbits -= 8;
    }
  }
  return new Uint8Array(out);
}

File: process-monitor/src/license/format.rs
//! Pretty license key format — the customer-facing encoding.
//!
//! Canonical (internal) form — what the Ed25519 signature covers:
//!     base64(JSON payload) "." base64(64-byte signature)
//!
//! Display (pretty) form — what customers receive and type:
//!     TALUS-XXXXX-XXXXX-…-XXXXX-SS
//!
//!   • XXXXX = Crockford base32 (0-9, A-Z minus I, L, O, U) of the canonical
//!     bytes, grouped in fives — no ambiguous characters.
//!   • SS    = last two hex chars of SHA-256 over the canonical bytes, so
//!     typos are caught before activation.
//!
//! The signature is PART of the encoded bytes, so the pretty form carries the
//! full signed key. Decoding is lossless. Old canonical keys stay valid.

use anyhow::{bail, Context, Result};
use sha2::{Digest, Sha256};

const ALPHABET: &[u8; 32] = b"0123456789ABCDEFGHJKMNPQRSTVWXYZ";
const PREFIX: &str = "TALUS";
#[allow(dead_code)] // license-issuer WIP — używane przez scripts/issuer-daemon (odpalamy po integracji)
const GROUP: usize = 5;

/// True when the string looks like the pretty display form (the full JWT
/// encoded — many groups). Short keys (TALUS-XXXXX-XXXXX-XXXXX-XXXXX) are
/// NOT pretty: they resolve to the JWT through the license server.
pub fn is_pretty(key: &str) -> bool {
    let compact: String = key.trim().chars().filter(|c| !c.is_whitespace()).collect();
    let upper = compact.to_uppercase();
    upper.starts_with(&format!("{PREFIX}-")) && upper.matches('-').count() >= 6
}

/// True when the string looks like a SHORT key (server-resolved JWT):
/// exactly TALUS-XXXXX-XXXXX-XXXXX-XXXXX — 4 groups of 5.
pub fn is_short(key: &str) -> bool {
    let compact: String = key.trim().chars().filter(|c| !c.is_whitespace()).collect();
    let upper = compact.to_uppercase();
    upper.starts_with(&format!("{PREFIX}-")) && upper.matches('-').count() == 4
}

/// Canonical key string → pretty key string.
#[allow(dead_code)] // license-issuer WIP — pretty-print dla kluczy w CLI issuer
pub fn to_pretty(canonical_key: &str) -> String {
    let data = canonical_key.trim().as_bytes();
    let digest = Sha256::digest(data);
    let body = b32_encode(data);
    let grouped: Vec<String> = body
        .as_bytes()
        .chunks(GROUP)
        .map(|c| String::from_utf8_lossy(c).into_owned())
        .collect();
    let checksum = hex_last2(&digest);
    format!("{PREFIX}-{}-{}", grouped.join("-"), checksum)
}

/// Pretty key string → canonical key string. Fails on checksum mismatch.
pub fn decode_pretty(pretty_key: &str) -> Result<String> {
    let cleaned: String = pretty_key
        .trim()
        .chars()
        .filter(|c| !c.is_whitespace())
        .collect::<String>()
        .to_uppercase();
    let rest = cleaned
        .strip_prefix(&format!("{PREFIX}-"))
        .ok_or_else(|| anyhow::anyhow!("not a pretty key"))?;
    let parts: Vec<&str> = rest.split('-').collect();
    if parts.len() < 2 {
        bail!("key too short");
    }
    let checksum = parts[parts.len() - 1];
    let body: String = parts[..parts.len() - 1].concat();
    for ch in body.chars() {
        if !ALPHABET.contains(&(ch as u8)) {
            bail!("invalid character '{ch}' in key");
        }
    }
    let data = b32_decode(&body)?;
    let digest = Sha256::digest(&data);
    let expected = hex_last2(&digest);
    if !checksum.eq_ignore_ascii_case(&expected) {
        bail!("checksum mismatch — retype the key");
    }
    String::from_utf8(data).context("decoded key is not valid UTF-8")
}

fn hex_last2(digest: &[u8]) -> String {
    let hex = digest.iter().map(|b| format!("{b:02x}")).collect::<String>();
    hex[hex.len() - 2..].to_uppercase()
}

#[allow(dead_code)] // license-issuer WIP — fallback kodowania bez base32 crate
fn b32_encode(data: &[u8]) -> String {
    let mut out = String::new();
    let mut acc: u32 = 0;
    let mut nbits: u32 = 0;
    for &byte in data {
        acc = (acc << 8) | u32::from(byte);
        nbits += 8;
        while nbits >= 5 {
            out.push(ALPHABET[((acc >> (nbits - 5)) & 31) as usize] as char);
            nbits -= 5;
        }
    }
    if nbits > 0 {
        out.push(ALPHABET[((acc << (5 - nbits)) & 31) as usize] as char);
    }
    out
}

fn b32_decode(s: &str) -> Result<Vec<u8>> {
    let mut out = Vec::new();
    let mut acc: u32 = 0;
    let mut nbits: u32 = 0;
    for ch in s.chars() {
        let idx = ALPHABET
            .iter()
            .position(|&c| c as char == ch)
            .ok_or_else(|| anyhow::anyhow!("invalid character"))?;
        acc = (acc << 5) | idx as u32;
        nbits += 5;
        if nbits >= 8 {
            out.push(((acc >> (nbits - 8)) & 0xff) as u8);
            nbits -= 8;
        }
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    const SAMPLE: &str = "eyJsaWNlbnNlX2lkIjoiVEFMVVMifQ.MEUCIQTest";

    #[test]
    fn short_key_classification() {
        assert!(is_short("talus-a2b3c-4d5e6-f7g8h-9j0k1"));
        assert!(is_short("  TALUS-A2B3C-4D5E6-F7G8H-9J0K1\n"));
        // Pretty full-JWT keys have many groups — not short.
        assert!(!is_short(&to_pretty(SAMPLE)));
        // Canonical has no TALUS prefix.
        assert!(!is_short(SAMPLE));
    }

    #[test]
    fn roundtrip() {
        assert!(is_pretty(&to_pretty(SAMPLE)));
        let decoded = decode_pretty(&to_pretty(SAMPLE)).unwrap();
        assert_eq!(decoded, SAMPLE);
    }

    #[test]
    fn checksum_catches_typos() {
        let pretty = to_pretty(SAMPLE);
        let corrupted = pretty.replace('A', "B");
        assert_ne!(corrupted, pretty);
        assert!(decode_pretty(&corrupted).is_err());
    }

    #[test]
    fn whitespace_and_case_tolerant() {
        let pretty = to_pretty(SAMPLE);
        let lowered = format!("  {}  ", pretty.to_lowercase()).replace('-', " - ");
        assert_eq!(decode_pretty(&lowered).unwrap(), SAMPLE);
    }
}
